website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md

```diff
@@ -0,0 +1,66 @@
+
+<!--truncate-->
+
+### OpenID Connect
+
+OpenID Connect (from here on "OIDC") is a standard that builds on top of OAuth 2.0, an existing standard. OIDC adds a standard for user identity on top of OAuth, which allows RPs (Relying Parties, more commonly referred to as "Clients") to verify the identity of users through tokens issued by an IDP (Identity providers). Before OIDC was as broadly used as it is now, most identity providers relied on a custom method for this, which made it much harder to implement broad support in clients. For example, a client would have to implement custom logic for logging in with Google, GitHub, Microsoft, etc. This also made it much harder for new identity providers (hey that's us!) to work with existing clients, as identity providers would have to basically emulate the specific implementation by one of the aforementioned providers.
+
+Since its introduction in [2014](<https://en.wikipedia.org/wiki/OpenID#OpenID_Connect_(OIDC)>), OIDC has been broadly adopted in the late 2010s, and now most providers support it (there's still a large list of providers that rely on pure OAuth with a custom identity layer on top, like Facebook, Twitter, Apple, but hopefully they'll migrate to OIDC eventually).
+
+The broad adoption on both sides has also increased the diversity and fragmentation, and indirectly also the diversion away from the standard.
+
+### Standards and testing
+
+One aspect where OIDC is very different to other standards is that they have a certification program, where applications can run through a (quite large) set of tests specified by the foundation itself, called [Conformance Tests](https://openid.net/certification/). This allows library authors for both clients and identity providers to ensure their implementation matches the standard, and behaves as expected. These tests are very thorough (the basic identity provider test suite contains 96 tests), which ensures that edge cases, uncommonly used features and both positive and negative flows are tested.
+
+These tests are all very well defined, they explain exactly what is tested, what the expected behavior is, and what (if any) possible issues are.
+
+For example, here you can see the overview of a test that's failed: ![Failed test overview](./failed-overview.png) and the exact request and reason why it failed: ![Failed test detail](./failed-detail.png)
+
+The exact standard definition is linked, which makes it very easy to dig in further and figure out what's supposed to happen.
+
+For authentik, we use these tests to ensure that we adhere to the standard as much as possible to increase compatibility as much as possible. We've got all tests passing on the current development builds (which actually helped us a couple of bugs), and we'll be certifying the identity provider portion of authentik very soon.
```

Comment:
will be certifying means getting an OpenID certification from the foundation for the OpenID Identity Provider component of authentik

And while I would love to run these tests automatically before every release, the tests require a lot of manual work so that's sadly not yet possible